usr/sbin/rear

```diff
@@ -510,41 +518,122 @@ fi
 # "(( EXIT_FAIL_MESSAGE )) && echo '${MESSAGE_PREFIX}$PROGRAM $WORKFLOW failed, check $RUNTIME_LOGFILE for details' 1>&8"
 # "exec 8>&-"
 # "exec 7>&-"
 # "exec 6<&-"
 
-# Used to determine whether TMPDIR has been changed in user config.
-# After the `mktemp` below, changing TMPDIR does not have the intended effect.
-# Save the current value to detect changes.
-saved_tmpdir="${TMPDIR-}"
+# Set RECOVERY_MODE when we are running inside the rescue/recovery system.
+# RECOVERY_MODE is a boolean variable (cf. conf/default.conf) because we need it below
+# to set TMPDIR to ReaR's TMP_DIR only when we are not running inside the recovery system
+# but we do not yet have lib/global-functions.sh (which contains is_false) sourced here.
+# In the recovery system /etc/rear-release is unique (it does not exist otherwise)
+# see build/default/970_add_rear_release.sh that adds it to identify the rescue environment:
+test -e "/etc/rear-release" && RECOVERY_MODE='y' || RECOVERY_MODE=''
+readonly RECOVERY_MODE
+
+# Create a sufficiently safe temporary working area BUILD_DIR for ReaR and
+# additionally set TMPDIR to TMP_DIR within BUILD_DIR for programs that are called by ReaR
+# (the latter only when ReaR runs on the original system i.e. when we are not in RECOVERY_MODE):
 if test "$WORKFLOW" != "help" ; then
-    # Prepend temporary working area (BUILD_DIR) removal exit task
-    # so it gets executed directly before the above listed exit tasks.
-    # This must be done before the first possible call of the 'Error' function
-    # otherwise we may error out but leave the build area behind:
+    # ReaR fails when TMPDIR is a relative directory, see
+    # https://github.com/rear/rear/pull/3168#issuecomment-1999996162
+    # so we set TMPDIR to its resolved absolute path.
+    # Ignore when 'readlink -e' fails (e.g. it won't fail when TMPDIR is a regular file)
+    # and verify afterwards that the 'readlink -e' result is a directory:
+    export TMPDIR="$( readlink -e "${TMPDIR-}" || true )"
+    if ! test -d "$TMPDIR" ; then
+        # Don't use Error() unless "QuietAddExitTask cleanup_build_area_and_end_program" was done:
+        echo "ERROR: TMPDIR '$TMPDIR' is no directory" >&2
+        exit 1
+    fi
+    # We rely on sufficiently safe 'mktemp -d' behaviour
+    # that the BUILD_DIR directory will have read, write, and search permissions
+    # for the current user (i.e. for root), but no permissions for the group or others
+    # (i.e. BUILD_DIR is "drwx------ root root /var/tmp/rear.XXXXXXXXXXXXXXX"), cf.
+    # https://www.gnu.org/software/coreutils/manual/html_node/mktemp-invocation.html#mktemp-invocation
+    # Create ReaR's working area BUILD_DIR:
+    if ! BUILD_DIR="$( mktemp -d -t rear.XXXXXXXXXXXXXXX )" ; then
+        # Don't use Error() unless "QuietAddExitTask cleanup_build_area_and_end_program" was done:
+        echo "ERROR: Could not create build area '$BUILD_DIR'" >&2
```

Comment:
when mktemp fails, the output will be empty and the message `ERROR: Could not create build area ''` a bit confusing. Of course this problem has already existed before so you don't have to fix it now.

Reply:
Fixed by
https://github.com/rear/rear/commit/96867121a3cdf0596d1f0d9ea0458009fa2703c3